src/Middleware/WebSockets/src/WebSocketMiddleware.cs

```diff
@@ -139,21 +144,77 @@ public async Task<WebSocket> AcceptAsync(WebSocketAcceptContext acceptContext)
                 }
 
                 TimeSpan keepAliveInterval = _options.KeepAliveInterval;
+                bool enableCompression = false;
+                bool serverContextTakeover = true;
+                int serverMaxWindowBits = 15;
                 if (acceptContext is ExtendedWebSocketAcceptContext advancedAcceptContext)
                 {
                     if (advancedAcceptContext.KeepAliveInterval.HasValue)
                     {
                         keepAliveInterval = advancedAcceptContext.KeepAliveInterval.Value;
                     }
+                    enableCompression = advancedAcceptContext.DangerousEnableCompression;
+                    serverContextTakeover = advancedAcceptContext.ServerContextTakeover;
+                    serverMaxWindowBits = advancedAcceptContext.ServerMaxWindowBits;
                 }
 
                 string key = _context.Request.Headers.SecWebSocketKey;
 
                 HandshakeHelpers.GenerateResponseHeaders(key, subProtocol, _context.Response.Headers);
 
+                WebSocketDeflateOptions? deflateOptions = null;
+                if (enableCompression)
+                {
+                    var ext = _context.Request.Headers.SecWebSocketExtensions;
+                    if (ext.Count != 0)
+                    {
+                        // loop over each extension offer, extensions can have multiple offers we can accept any
+                        foreach (var extension in _context.Request.Headers.GetCommaSeparatedValues(HeaderNames.SecWebSocketExtensions))
+                        {
+                            if (extension.TrimStart().StartsWith("permessage-deflate", StringComparison.Ordinal))
+                            {
+                                deflateOptions = new WebSocketDeflateOptions()
+                                {
+                                    ServerContextTakeover = serverContextTakeover,
+                                    ServerMaxWindowBits = serverMaxWindowBits
+                                };
+                                if (HandshakeHelpers.ParseDeflateOptions(extension, deflateOptions, out var response))
```

Comment:
deflateOptions is still set even if ParseDeflateOptions returns false, which would result in the server using compression but the client not knowing it.

Change deflateOptions to an out paramter and pass in the serverContextTakeover and serverMaxWindowBits so deflateOptions only gets set if ParseDeflateOptions returns true.